README.md
# adaptive-motion v0.2.0 [![Build Status](https://travis-ci.org/angular-adaptive/adaptive-motion.png?branch=master)](https://travis-ci.org/angular-adaptive/adaptive-motion)

This module allows you to control an AngularJS app using web camera.

### Demo

Check out http://angular-adaptive.github.io/adaptive-motion/demo/

# Requirements

- AngularJS v 1.0+
- [getUserMedia Stream](http://caniuse.com/#feat=stream) support

# Usage

We use [bower](http://twitter.github.com/bower/) for dependency management. Add

    dependencies: {
        "angular-adaptive-motion": "latest"
    }

To your `bower.json` file. Then run

    bower install

This will copy the angular-adaptive-motion files into your `bower_components` folder, along with its dependencies. Load the script files in your application:

    <script type="text/javascript" src="bower_components/angular/angular.js"></script>
    <script type="text/javascript" src="bower_components/angular-adaptive-motion/angular-adaptive-motion.js"></script>

Add the **adaptive.motion** module as a dependency to your application module:

    var myAppModule = angular.module('MyApp', ['adaptive.motion']);

and include `$motion` service as a dependency to your controller:

    angular.module('MyApp').controller('MainCtrl', function ['$scope', '$motion', ($scope, $motion) {

    }]);

### Public methods

#### $motion.start();
Starts gesture recognition.

#### $motion.stop();
Stops gesture recognition.

#### $motion.onStart(cb);
On start callback.

#### $motion.onStop(cb);
On stop callback.

#### $motion.onError(cb);
On error callback.

#### $motion.onSwipeLeft(cb);
On swipe left gesture.
```
$motion.onSwipeLeft(function(data){
    $scope.$apply(function(){
        console.log('onSwipeLeft');
    });
});
```

#### $motion.onSwipeRight(cb);
On swipe right gesture.
```
$motion.onSwipeRight(function(data){
    $scope.$apply(function(){
        console.log('onSwipeRight');
    });
});
```

#### $motion.onSwipeUp(cb);
On swipe up gesture.
```
$motion.onSwipeUp(function(data){
    $scope.$apply(function(){
        console.log('onSwipeUp');
    });
});
```

#### $motion.onSwipeDown(cb);
On swipe down gesture.
```
$motion.onSwipeDown(function(data){
    $scope.$apply(function(){
        console.log('onSwipeDown');
    });
});
```


### Configuration

You can configure `$motionProvider` to a custom treshold options in app configuration.

```
$motionProvider.setTreshold({
    'rgb': 150,
    'move': 3,
    'bright': 300
});
```

You can also set custom hsv filter.

```
$motionProvider.setHsvFilter({
    'huemin': 0.0,
    'huemax': 0.1,
    'satmin': 0.0,
    'satmax': 1.0,
    'valmin': 0.4,
    'valmax': 1.0
});
```

### Visualization

If you want to visualize you can add `adaptive-motion` attribute into your canvas element. 
You can choose from following styles:


#### Video

```
<canvas adaptive-motion="video"></canvas>
```

![video](https://raw.github.com/angular-adaptive/adaptive-motion/canary/screens/video.png)

#### Skin

```
<canvas adaptive-motion="skin"></canvas>
```

![skin](https://raw.github.com/angular-adaptive/adaptive-motion/canary/screens/skin.png)

#### Edge

```
<canvas adaptive-motion="edge"></canvas>
```

![edge](https://raw.github.com/angular-adaptive/adaptive-motion/canary/screens/edges.png)
    
# Contributing

Contributions are welcome. Please make a pull request against canary branch and do not bump versions. Also include tests.

# Testing

We use karma and jshint to ensure the quality of the code. The easiest way to run these checks is to use grunt:

    npm install -g grunt-cli
    npm install
    bower install
    grunt

The karma task will try to open Chrome as a browser in which to run the tests. Make sure this is available or change the configuration in `test/test.config.js` 

# References

Check out [willy-vvu`s](https://github.com/willy-vvu) webcam-based control of [reveal.js fork](https://github.com/willy-vvu/reveal.js).


# License

The MIT License

Copyright (c) 2013 Jan Antala, http://janantala.com


CHANGELOG.md
<a name="v0.2.0"></a>
## v0.2.0 (2013-11-12)


#### Features

* **GruntFile:** add concated and uglified version ([92fecaed](https://github.com/angular-adaptive/adaptive-motion/commit/92fecaed4f0365b8987a0a848921447081bdc349))
* **conventional-changelog:** adds grunt-conventional-changelog ([432cf774](https://github.com/angular-adaptive/adaptive-motion/commit/432cf7742466b7152250869b930b9627e1e53661))


#### Breaking Changes

* src/adaptive-motion.js is now renamed into src/angular-adaptive-motion.js
 ([92fecaed](https://github.com/angular-adaptive/adaptive-motion/commit/92fecaed4f0365b8987a0a848921447081bdc349))



test/test.conf.js
module.exports = function(config) {
  config.set({
    basePath: '..',
    files: [
      'bower_components/angular/angular.js',
      'bower_components/angular-mocks/angular-mocks.js',
      'src/angular-adaptive-motion.js',
      'test/*.spec.js'
    ],
    frameworks: ['jasmine'],
    singleRun: true,
    browsers: [ 'Chrome' ]
  });
};

test/unit.spec.js
describe('adaptive.motion', function() {

  describe('$motionProvider config', function() {
    var $motion;

    beforeEach(module('adaptive.motion', function($motionProvider) {
      $motion = $motionProvider;
    }));

    it('should be defined', inject(function() {
      expect($motion).toBeDefined();
      expect(typeof $motion).toBe('object');
    }));

    it('should have default treshold options', inject(function() {
      expect($motion.treshold).toBeDefined();
      expect($motion.treshold.rgb).toBeDefined();
      expect($motion.treshold.move).toBeDefined();
      expect($motion.treshold.bright).toBeDefined();
    }));

    it('should have setTreshold method', inject(function() {
      expect($motion.setTreshold).toBeDefined();
      expect(typeof $motion.setTreshold).toBe('function');

      $motion.setTreshold({
        'rgb': 123
      });

      expect($motion.treshold.rgb).toBe(123);
    }));

  });

  describe('$get methods', function() {

    beforeEach(module('adaptive.motion', function() {

    }));

    it('should have public methods', inject(function($motion) {
      expect($motion).toBeDefined();
      expect($motion.start).toBeDefined();
      expect($motion.stop).toBeDefined();
      expect($motion.onStart).toBeDefined();
      expect($motion.onStop).toBeDefined();
      expect($motion.onError).toBeDefined();
      expect($motion.onSwipeLeft).toBeDefined();
      expect($motion.onSwipeRight).toBeDefined();
      expect($motion.onSwipeUp).toBeDefined();
      expect($motion.onSwipeDown).toBeDefined();

      expect(typeof $motion.start).toBe('function');
      expect(typeof $motion.stop).toBe('function');
      expect(typeof $motion.onStart).toBe('function');
      expect(typeof $motion.onStop).toBe('function');
      expect(typeof $motion.onError).toBe('function');
      expect(typeof $motion.onSwipeLeft).toBe('function');
      expect(typeof $motion.onSwipeRight).toBe('function');
      expect(typeof $motion.onSwipeUp).toBe('function');
      expect(typeof $motion.onSwipeDown).toBe('function');
    }));

    it('should have called rootScope.$broadcast', inject(function($motion, $rootScope){

      spyOn($rootScope, '$on').andCallThrough();
      spyOn($rootScope, '$broadcast').andCallThrough();
      $rootScope.$apply();

      var counter = 0;

      $motion.onSwipeLeft(function(){
        counter++;
      });

      $motion.onSwipeRight(function(){
        counter++;
      });

      $motion.onSwipeUp(function(){
        counter++;
      });

      $motion.onSwipeDown(function(){
        counter++;
      });
      

      $rootScope.$broadcast('adaptive.motion:onSwipeLeft');
      expect($rootScope.$broadcast).toHaveBeenCalled();
      expect(counter).toBe(1);

      $rootScope.$broadcast('adaptive.motion:onSwipeRight');
      expect($rootScope.$broadcast).toHaveBeenCalled();
      expect(counter).toBe(2);

      $rootScope.$broadcast('adaptive.motion:onSwipeUp');
      expect($rootScope.$broadcast).toHaveBeenCalled();
      expect(counter).toBe(3);

      $rootScope.$broadcast('adaptive.motion:onSwipeDown');
      expect($rootScope.$broadcast).toHaveBeenCalled();
      expect(counter).toBe(4);
    }));

  });
});

src/angular-adaptive-motion.js
(function () {
'use strict';

var adaptive = angular.module('adaptive.motion', []);

// RequestAnimationFrame fallback
(function() {
  var lastTime = 0;
  var vendors = ['webkit', 'moz'];
  for(var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
      window.requestAnimationFrame = window[vendors[x]+'RequestAnimationFrame'];
      window.cancelAnimationFrame =
        window[vendors[x]+'CancelAnimationFrame'] || window[vendors[x]+'CancelRequestAnimationFrame'];
  }

  if (!window.requestAnimationFrame) {
      window.requestAnimationFrame = function(callback, element) {
          var currTime = new Date().getTime();
          var timeToCall = Math.max(0, 16 - (currTime - lastTime));
          var id = window.setTimeout(function() { callback(currTime + timeToCall); },
            timeToCall);
          lastTime = currTime + timeToCall;
          return id;
      };
  }

  if (!window.cancelAnimationFrame) {
      window.cancelAnimationFrame = function(id) {
          clearTimeout(id);
      };
  }
}());

/**
 * Converts rgb into hsv
 * @param  {Integer} r
 * @param  {Integer} g
 * @param  {Integer} b
 * @return {Array}
 */
var rgb2Hsv = function(r, g, b){

  r = r/255;
  g = g/255;
  b = b/255;

  var max = Math.max(r, g, b);
  var min = Math.min(r, g, b);

  var h, s, v = max;

  var d = max - min;

  s = max === 0 ? 0 : d / max;

  if (max === min){
      h = 0; // achromatic
  }
  else{
    switch(max){
      case r: h = (g - b) / d + (g < b ? 6 : 0); break;
      case g: h = (b - r) / d + 2; break;
      case b: h = (r - g) / d + 4; break;
    }
    h /= 6;
  }

  return [h, s, v];
};

adaptive.provider('$motion', [function() {

  var requestId;
  var video = document.createElement('video');
  video.setAttribute('autoplay', 'true');
  video.setAttribute('width', '300');
  var canvas = document.createElement('canvas');
  var context = canvas.getContext('2d');

  this.treshold = {
    'rgb': 150,
    'move': 2,
    'bright': 300
  };

  this.hsvFilter = {
    'huemin': 0.0,
    'huemax': 0.1,
    'satmin': 0.0,
    'satmax': 1.0,
    'valmin': 0.4,
    'valmax': 1.0
  };

  this.setTreshold = function(treshold) {
    angular.extend(this.treshold, treshold);
  };

  this.setHsvFilter = function(hsvFilter) {
    angular.extend(this.hsvFilter, hsvFilter);
  };

  this.$get = function($rootScope) {

    var treshold = this.treshold;
    var hsvFilter = this.hsvFilter;

    var compression = 5;
    var width = 0;
    var height = 0;

    var draw;
    var localMediaStream;

    var lastDraw;
    var lastDown = {
      x: 0,
      y: 0,
      d: 0
    };

    var start = function(){

      window.URL = window.URL || window.webkitURL;
      navigator.getUserMedia = navigator.getUserMedia || navigator.webkitGetUserMedia || navigator.mozGetUserMedia || navigator.msGetUserMedia;

      if (navigator.getUserMedia) {
        navigator.getUserMedia({audio: false, video: true},
          function(stream){
            $rootScope.$broadcast('adaptive.motion:onStart');
            localMediaStream = stream;
            video.src = window.URL.createObjectURL(stream);
            video.addEventListener('play', function() {
              requestId = window.requestAnimationFrame(redraw);
            });
          },
          function(){
            $rootScope.$broadcast('adaptive.motion:onError', 'Access denied!');
            throw new Error('Access denied!');
          }
        );
      }
      else {
        $rootScope.$broadcast('adaptive.motion:onError', 'getUserMedia() is not supported in your browser');
        throw new Error('getUserMedia() is not supported in your browser');
      }
    };

    var stop = function(){
      window.cancelAnimationFrame(requestId);
      if (localMediaStream) {
        localMediaStream.stop();
      }
      localMediaStream = undefined;
      $rootScope.$broadcast('adaptive.motion:onStop');
    };

    var redraw = function() {
      if (canvas.width !== video.videoWidth){
        width = Math.floor(video.videoWidth / compression);
        height = Math.floor(video.videoHeight / compression);
        canvas.width = width;
        canvas.height = height;
      }

      try {
        context.drawImage(video,0,0,width,height);
        draw = context.getImageData(0, 0, width, height);
        $rootScope.$broadcast('adaptive.motion:videoData', draw);
        var skinFilter = filterSkin(draw);
        lastDraw = getMovements(skinFilter);

        requestId = window.requestAnimationFrame(redraw);
      }
      catch (e) {
        if (e.name === 'NScontextERRORcontextNOTcontextAVAILABLE') {
          requestId = window.requestAnimationFrame(redraw);
        }
        else {
          throw e;
        }
      }
    };

    /**
     * Filters skin from video image data
     * @param  {ImageData} video
     * @return {ImageData}
     */
    var filterSkin = function(video){

      var skinFilter = context.getImageData(0,0,width,height);
      var totalPixels = skinFilter.width * skinFilter.height;
      var index = totalPixels * 4;

      var pix = 0;
      for (var y=0; y<height; y++)
      {
        for (var x=0; x<width; x++)
        {
          index = x+y*width;
          var r = video.data[pix];
          var g = video.data[pix+1];
          var b = video.data[pix+2];
          var a = video.data[pix+3];

          var hsv = rgb2Hsv(r,g,b);
          //When the hand is too lose (hsv[0] > 0.59 && hsv[0] < 1.0)
          //Skin Range on HSV values
          if(((hsv[0] > hsvFilter.huemin && hsv[0] < hsvFilter.huemax)||(hsv[0] > 0.59 && hsv[0] < 1.0))&&(hsv[1] > hsvFilter.satmin && hsv[1] < hsvFilter.satmax)&&(hsv[2] > hsvFilter.valmin && hsv[2] < hsvFilter.valmax)){
            skinFilter[pix] = r;
            skinFilter[pix+1] = g;
            skinFilter[pix+2] = b;
            skinFilter[pix+3] = a;
          }
          else{
            skinFilter.data[pix] = 255;
            skinFilter.data[pix+1] = 255;
            skinFilter.data[pix+2] = 255;
            skinFilter.data[pix+3] = 255;
          }

          pix = index * 4;
        }
      }
      $rootScope.$broadcast('adaptive.motion:skinData', skinFilter);
      return skinFilter;
    };

    /**
     * Gets movement data
     * @param  {ImageData} draw
     * @return {ImageData}
     */
    var getMovements = function(draw){
      var edge = context.createImageData(width, height);
      var totalx = 0;
      var totaly = 0;
      var changed = 0;
      var pix = edge.width * edge.height * 4;

      if (lastDraw){

        while ((pix -= 4) > 0) {
          var rgbaDelta = Math.abs(draw.data[pix] - lastDraw.data[pix]) +
                  Math.abs(draw.data[pix+1] - lastDraw.data[pix+1]) +
                  Math.abs(draw.data[pix+2] - lastDraw.data[pix+2]);

          if (rgbaDelta > treshold.rgb){
            edge.data[pix] = 0;
            edge.data[pix+1] = 0;
            edge.data[pix+2] = 0;
            edge.data[pix+3] = 255;
            changed += 1;
            totalx += (pix/4) % width;
            totaly += Math.floor((pix/4) / edge.height);
          }
          else {
            edge.data[pix] = 255;
            edge.data[pix+1] = 255;
            edge.data[pix+2] = 255;
            edge.data[pix+3] = 255;
          }
        }
      }

      if (changed){
        $rootScope.$broadcast('adaptive.motion:edgeData', edge);

        var down = {
          x: totalx / changed,
          y: totaly / changed,
          d: changed
        };
        recognizeGesture(down);
      }

      return draw;
    };

    /**
     * Sets last down
     * @param {Object} down
     */
    var setLastDown = function(down){
      lastDown = {
        x: down.x,
        y: down.y,
        d: down.d
      };
    };

    var avg = 0;
    var state = 0; //States: 0 waiting for gesture, 1 waiting for next move after gesture, 2 waiting for gesture to end

    /**
     * Recognizes gesture
     * @param  {Object} down
     */
    var recognizeGesture = function(down){
      avg = 0.9 * avg + 0.1 * down.d;
      var davg = down.d - avg;
      var foundGesture = davg > treshold.bright;

      switch (state){
        case 0:
          if (foundGesture){ //Found a gesture, waiting for next move
            state = 1;
            setLastDown(down);
          }
          break;
        case 2: //Wait for gesture to end
          if (!foundGesture){ //Gesture ended
            state = 0;
          }
          break;
        case 1: //Got next move, do something based on direction
          var dx = down.x - lastDown.x;
          var dy = down.y - lastDown.y;
          var dirx = Math.abs(dy) < Math.abs(dx) - treshold.move;
          var diry = Math.abs(dx) < Math.abs(dy) - treshold.move;
          // console.log(dx, dy, dirx);

          if (dirx) {
            if (dx < - treshold.move){
              $rootScope.$broadcast('adaptive.motion:onSwipeRight');
            }
            else if (dx > treshold.move){
              $rootScope.$broadcast('adaptive.motion:onSwipeLeft');
            }
          }
          else if (diry) {
            if (dy > treshold.move){
              $rootScope.$broadcast('adaptive.motion:onSwipeDown');
            }
            else if (dy < - treshold.move){
              $rootScope.$broadcast('adaptive.motion:onSwipeUp');
            }
          }

          state = 2;
          break;
      }
    };

    var onStart = function(cb){
      $rootScope.$on('adaptive.motion:onStart', function(e, data){
        cb(data);
      });
    };

    var onStop = function(cb){
      $rootScope.$on('adaptive.motion:onStop', function(e, data){
        cb(data);
      });
    };

    var onError = function(cb){
      $rootScope.$on('adaptive.motion:onError', function(e, data){
        cb(data);
      });
    };

    var onSwipeLeft = function(cb){
      $rootScope.$on('adaptive.motion:onSwipeLeft', function(e, data){
        cb(data);
      });
    };

    var onSwipeRight = function(cb){
      $rootScope.$on('adaptive.motion:onSwipeRight', function(e, data){
        cb(data);
      });
    };

    var onSwipeUp = function(cb){
      $rootScope.$on('adaptive.motion:onSwipeUp', function(e, data){
        cb(data);
      });
    };

    var onSwipeDown = function(cb){
      $rootScope.$on('adaptive.motion:onSwipeDown', function(e, data){
        cb(data);
      });
    };

    return {
      start: function(){
        start();
      },
      stop: function(){
        stop();
      },
      onStart: function(cb){
        onStart(cb);
      },
      onStop: function(cb){
        onStop(cb);
      },
      onError: function(cb){
        onError(cb);
      },
      onSwipeLeft: function(cb){
        onSwipeLeft(cb);
      },
      onSwipeRight: function(cb){
        onSwipeRight(cb);
      },
      onSwipeUp: function(cb){
        onSwipeUp(cb);
      },
      onSwipeDown: function(cb){
        onSwipeDown(cb);
      }
    };
  };
}]);

adaptive.directive('adaptiveMotion', ['$rootScope', function ($rootScope) {
  return {
    restrict: 'A',
    link: function postLink(scope, element, attrs) {
      var canvas = element[0];
      var context = canvas.getContext('2d');

      if (attrs['adaptiveMotion'] === 'video'){
        $rootScope.$on('adaptive.motion:videoData', function(e, data){
          context.putImageData(data, 0, 0);
        });
      }
      else if (attrs['adaptiveMotion'] === 'skin'){
        $rootScope.$on('adaptive.motion:skinData', function(e, data){
          context.putImageData(data, 0, 0);
        });
      }
      else {
        $rootScope.$on('adaptive.motion:edgeData', function(e, data){
          context.putImageData(data, 0, 0);
        });
      }
    }
  };
}]);

})();

demo/js/controllers/mainCtrl.js
/*global motion */
(function () {
'use strict';

/**
 * The main controller for the app. 
 */
motion.controller('mainCtrl', function mainCtrl($scope, $rootScope, $motion) {

  $scope.visualization = 'edge';

  var playSound = function(){
    var audio = new Audio('/assets/lightsaber.wav');
    audio.play();
  };

  var addGesture = function(gesture){
    $scope.gestures.push(gesture);
  };

  $scope.gestures = [];
  $motion.start();

  $motion.onSwipeLeft(function(data){
    $scope.$apply(function(){
      console.log('onSwipeLeft');
      addGesture('Swipe Left');
      playSound();
    });
  });

  $motion.onSwipeRight(function(data){
    $scope.$apply(function(){
      console.log('onSwipeRight');
      addGesture('Swipe Right');
      playSound();
    });
  });

  $motion.onSwipeUp(function(data){
    $scope.$apply(function(){
      console.log('onSwipeUp');
      addGesture('Swipe Up');
      playSound();
    });
  });

  $motion.onSwipeDown(function(data){
    $scope.$apply(function(){
      console.log('onSwipeDown');
      addGesture('Swipe Down');
      playSound();
    });
  });

  $motion.onStart(function(){
    console.log('onStart');
  });

  $motion.onStop(function(){
    console.log('onStart');
  });

  $motion.onError(function(err){
    console.error('onError', err);
    alert(err);
  });

});

})();

demo/components/angular-adaptive/adaptive-motion/src/adaptive-motion.js
/**
 * angular-adaptive-motion v0.1.0
 * The MIT License
 * Copyright (c) 2013 Jan Antala http://janantala.com
 */

(function () {
'use strict';

var adaptive = angular.module('adaptive.motion', []);

// RequestAnimationFrame fallback
(function() {
  var lastTime = 0;
  var vendors = ['webkit', 'moz'];
  for(var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
      window.requestAnimationFrame = window[vendors[x]+'RequestAnimationFrame'];
      window.cancelAnimationFrame =
        window[vendors[x]+'CancelAnimationFrame'] || window[vendors[x]+'CancelRequestAnimationFrame'];
  }

  if (!window.requestAnimationFrame) {
      window.requestAnimationFrame = function(callback, element) {
          var currTime = new Date().getTime();
          var timeToCall = Math.max(0, 16 - (currTime - lastTime));
          var id = window.setTimeout(function() { callback(currTime + timeToCall); },
            timeToCall);
          lastTime = currTime + timeToCall;
          return id;
      };
  }

  if (!window.cancelAnimationFrame) {
      window.cancelAnimationFrame = function(id) {
          clearTimeout(id);
      };
  }
}());

/**
 * Converts rgb into hsv
 * @param  {Integer} r
 * @param  {Integer} g
 * @param  {Integer} b
 * @return {Array}
 */
var rgb2Hsv = function(r, g, b){

  r = r/255;
  g = g/255;
  b = b/255;

  var max = Math.max(r, g, b);
  var min = Math.min(r, g, b);

  var h, s, v = max;

  var d = max - min;

  s = max === 0 ? 0 : d / max;

  if (max === min){
      h = 0; // achromatic
  }
  else{
    switch(max){
      case r: h = (g - b) / d + (g < b ? 6 : 0); break;
      case g: h = (b - r) / d + 2; break;
      case b: h = (r - g) / d + 4; break;
    }
    h /= 6;
  }

  return [h, s, v];
};

adaptive.provider('$motion', [function() {

  var requestId;
  var video = document.createElement('video');
  video.setAttribute('autoplay', 'true');
  video.setAttribute('width', '300');
  var canvas = document.createElement('canvas');
  var context = canvas.getContext('2d');

  this.treshold = {
    'rgb': 150,
    'move': 2,
    'bright': 300
  };

  this.hsvFilter = {
    'huemin': 0.0,
    'huemax': 0.1,
    'satmin': 0.0,
    'satmax': 1.0,
    'valmin': 0.4,
    'valmax': 1.0
  };

  this.setTreshold = function(treshold) {
    angular.extend(this.treshold, treshold);
  };

  this.setHsvFilter = function(hsvFilter) {
    angular.extend(this.hsvFilter, hsvFilter);
  };

  this.$get = function($rootScope) {

    var treshold = this.treshold;
    var hsvFilter = this.hsvFilter;

    var compression = 5;
    var width = 0;
    var height = 0;

    var draw;
    var localMediaStream;

    var lastDraw;
    var lastDown = {
      x: 0,
      y: 0,
      d: 0
    };

    var start = function(){

      window.URL = window.URL || window.webkitURL;
      navigator.getUserMedia = navigator.getUserMedia || navigator.webkitGetUserMedia || navigator.mozGetUserMedia || navigator.msGetUserMedia;

      if (navigator.getUserMedia) {
        navigator.getUserMedia({audio: false, video: true},
          function(stream){
            $rootScope.$broadcast('adaptive.motion:onStart');
            localMediaStream = stream;
            video.src = window.URL.createObjectURL(stream);
            video.addEventListener('play', function() {
              requestId = window.requestAnimationFrame(redraw);
            });
          },
          function(){
            $rootScope.$broadcast('adaptive.motion:onError', 'Access denied!');
            throw new Error('Access denied!');
          }
        );
      }
      else {
        $rootScope.$broadcast('adaptive.motion:onError', 'getUserMedia() is not supported in your browser');
        throw new Error('getUserMedia() is not supported in your browser');
      }
    };

    var stop = function(){
      window.cancelAnimationFrame(requestId);
      if (localMediaStream) {
        localMediaStream.stop();
      }
      localMediaStream = undefined;
      $rootScope.$broadcast('adaptive.motion:onStop');
    };

    var redraw = function() {
      if (canvas.width !== video.videoWidth){
        width = Math.floor(video.videoWidth / compression);
        height = Math.floor(video.videoHeight / compression);
        canvas.width = width;
        canvas.height = height;
      }

      try {
        context.drawImage(video,0,0,width,height);
        draw = context.getImageData(0, 0, width, height);
        $rootScope.$broadcast('adaptive.motion:videoData', draw);
        var skinFilter = filterSkin(draw);
        lastDraw = getMovements(skinFilter);

        requestId = window.requestAnimationFrame(redraw);
      }
      catch (e) {
        if (e.name === 'NScontextERRORcontextNOTcontextAVAILABLE') {
          requestId = window.requestAnimationFrame(redraw);
        }
        else {
          throw e;
        }
      }
    };

    /**
     * Filters skin from video image data
     * @param  {ImageData} video
     * @return {ImageData}
     */
    var filterSkin = function(video){

      var skinFilter = context.getImageData(0,0,width,height);
      var totalPixels = skinFilter.width * skinFilter.height;
      var index = totalPixels * 4;

      var pix = 0;
      for (var y=0; y<height; y++)
      {
        for (var x=0; x<width; x++)
        {
          index = x+y*width;
          var r = video.data[pix];
          var g = video.data[pix+1];
          var b = video.data[pix+2];
          var a = video.data[pix+3];

          var hsv = rgb2Hsv(r,g,b);
          //When the hand is too lose (hsv[0] > 0.59 && hsv[0] < 1.0)
          //Skin Range on HSV values
          if(((hsv[0] > hsvFilter.huemin && hsv[0] < hsvFilter.huemax)||(hsv[0] > 0.59 && hsv[0] < 1.0))&&(hsv[1] > hsvFilter.satmin && hsv[1] < hsvFilter.satmax)&&(hsv[2] > hsvFilter.valmin && hsv[2] < hsvFilter.valmax)){
            skinFilter[pix] = r;
            skinFilter[pix+1] = g;
            skinFilter[pix+2] = b;
            skinFilter[pix+3] = a;
          }
          else{
            skinFilter.data[pix] = 255;
            skinFilter.data[pix+1] = 255;
            skinFilter.data[pix+2] = 255;
            skinFilter.data[pix+3] = 255;
          }

          pix = index * 4;
        }
      }
      $rootScope.$broadcast('adaptive.motion:skinData', skinFilter);
      return skinFilter;
    };

    /**
     * Gets movement data
     * @param  {ImageData} draw
     * @return {ImageData}
     */
    var getMovements = function(draw){
      var edge = context.createImageData(width, height);
      var totalx = 0;
      var totaly = 0;
      var changed = 0;
      var pix = edge.width * edge.height * 4;

      if (lastDraw){

        while ((pix -= 4) > 0) {
          var rgbaDelta = Math.abs(draw.data[pix] - lastDraw.data[pix]) +
                  Math.abs(draw.data[pix+1] - lastDraw.data[pix+1]) +
                  Math.abs(draw.data[pix+2] - lastDraw.data[pix+2]);

          if (rgbaDelta > treshold.rgb){
            edge.data[pix] = 0;
            edge.data[pix+1] = 0;
            edge.data[pix+2] = 0;
            edge.data[pix+3] = 255;
            changed += 1;
            totalx += (pix/4) % width;
            totaly += Math.floor((pix/4) / edge.height);
          }
          else {
            edge.data[pix] = 255;
            edge.data[pix+1] = 255;
            edge.data[pix+2] = 255;
            edge.data[pix+3] = 255;
          }
        }
      }

      if (changed){
        $rootScope.$broadcast('adaptive.motion:edgeData', edge);

        var down = {
          x: totalx / changed,
          y: totaly / changed,
          d: changed
        };
        recognizeGesture(down);
      }

      return draw;
    };

    /**
     * Sets last down
     * @param {Object} down
     */
    var setLastDown = function(down){
      lastDown = {
        x: down.x,
        y: down.y,
        d: down.d
      };
    };

    var avg = 0;
    var state = 0; //States: 0 waiting for gesture, 1 waiting for next move after gesture, 2 waiting for gesture to end

    /**
     * Recognizes gesture
     * @param  {Object} down
     */
    var recognizeGesture = function(down){
      avg = 0.9 * avg + 0.1 * down.d;
      var davg = down.d - avg;
      var foundGesture = davg > treshold.bright;

      switch (state){
        case 0:
          if (foundGesture){ //Found a gesture, waiting for next move
            state = 1;
            setLastDown(down);
          }
          break;
        case 2: //Wait for gesture to end
          if (!foundGesture){ //Gesture ended
            state = 0;
          }
          break;
        case 1: //Got next move, do something based on direction
          var dx = down.x - lastDown.x;
          var dy = down.y - lastDown.y;
          var dirx = Math.abs(dy) < Math.abs(dx) - treshold.move;
          var diry = Math.abs(dx) < Math.abs(dy) - treshold.move;
          // console.log(dx, dy, dirx);

          if (dirx) {
            if (dx < - treshold.move){
              $rootScope.$broadcast('adaptive.motion:onSwipeRight');
            }
            else if (dx > treshold.move){
              $rootScope.$broadcast('adaptive.motion:onSwipeLeft');
            }
          }
          else if (diry) {
            if (dy > treshold.move){
              $rootScope.$broadcast('adaptive.motion:onSwipeDown');
            }
            else if (dy < - treshold.move){
              $rootScope.$broadcast('adaptive.motion:onSwipeUp');
            }
          }

          state = 2;
          break;
      }
    };

    var onStart = function(cb){
      $rootScope.$on('adaptive.motion:onStart', function(e, data){
        cb(data);
      });
    };

    var onStop = function(cb){
      $rootScope.$on('adaptive.motion:onStop', function(e, data){
        cb(data);
      });
    };

    var onError = function(cb){
      $rootScope.$on('adaptive.motion:onError', function(e, data){
        cb(data);
      });
    };

    var onSwipeLeft = function(cb){
      $rootScope.$on('adaptive.motion:onSwipeLeft', function(e, data){
        cb(data);
      });
    };

    var onSwipeRight = function(cb){
      $rootScope.$on('adaptive.motion:onSwipeRight', function(e, data){
        cb(data);
      });
    };

    var onSwipeUp = function(cb){
      $rootScope.$on('adaptive.motion:onSwipeUp', function(e, data){
        cb(data);
      });
    };

    var onSwipeDown = function(cb){
      $rootScope.$on('adaptive.motion:onSwipeDown', function(e, data){
        cb(data);
      });
    };

    return {
      start: function(){
        start();
      },
      stop: function(){
        stop();
      },
      onStart: function(cb){
        onStart(cb);
      },
      onStop: function(cb){
        onStop(cb);
      },
      onError: function(cb){
        onError(cb);
      },
      onSwipeLeft: function(cb){
        onSwipeLeft(cb);
      },
      onSwipeRight: function(cb){
        onSwipeRight(cb);
      },
      onSwipeUp: function(cb){
        onSwipeUp(cb);
      },
      onSwipeDown: function(cb){
        onSwipeDown(cb);
      }
    };
  };
}]);

adaptive.directive('adaptiveMotion', ['$rootScope', function ($rootScope) {
  return {
    restrict: 'A',
    link: function postLink(scope, element, attrs) {
      var canvas = element[0];
      var context = canvas.getContext('2d');

      if (attrs['adaptiveMotion'] === 'video'){
        $rootScope.$on('adaptive.motion:videoData', function(e, data){
          context.putImageData(data, 0, 0);
        });
      }
      else if (attrs['adaptiveMotion'] === 'skin'){
        $rootScope.$on('adaptive.motion:skinData', function(e, data){
          context.putImageData(data, 0, 0);
        });
      }
      else {
        $rootScope.$on('adaptive.motion:edgeData', function(e, data){
          context.putImageData(data, 0, 0);
        });
      }
    }
  };
}]);

})();